### PR TITLE
NL: Provide feedback on various fallbacks

### PR DESCRIPTION
**Pull request by pradh**
Types of fallbacks:

1. Place and child-place-type fallbacks due to lack of data
2. Context fallbacks when we fail to recognize SVs or places mentioned in a query
3. Place Overview fallbacks when we have no SV or when SV has no data 

Also, avoid looking up contained-in place-type from past context (its very hard to reason).

Implementation wise see `maybe_add_fallback()` function, and `base.py` in general.

TODO:  For place fallbacks, instead of always display, expect user to click to show charts.

#### Examples for place + place-type fallbacks

image

image

image

#### Examples for context fallbacks

image

image

#### Examples for overview fallback

image

image

### CODE REVIEW

File: server/lib/nl/common/utterance.py [hunk 1]
```diff
@@ -254,6 +278,44 @@ def _dict_to_chart_spec(charts_dict: List[Dict]) -> List[ChartSpec]:
   return charts
 
 
+def _place_fallback_to_dict(pfb: PlaceFallback) -> Dict:
+  if not pfb:
+    return {}
+  pfb_dict = {}
+  if pfb.origType:
+    pfb_dict['origType'] = pfb.origType.value
```
Comment (jehangiramjad) [inline server/lib/nl/common/utterance.py:286]: nit: should the key strings be constants somewhere?

File: server/lib/nl/fulfillment/base.py [hunk 2]
```diff
@@ -374,9 +399,81 @@ def get_default_contained_in_place(state: PopulateState) -> Place:
...
+        # This is a perfectly legitimate case that happens
+        # when promoting SIMPLE to CONTAINED_IN for example.
+        # In this case, skip type matching.
+        orig_type = new_type
+      else:
+        # We are falling back to parent without a sub-type,
+        # so make it clear that we are no longer talking about
+        # "places in california" but "california" as a state.
+        orig_type = ContainedInPlaceType.PLACE
+    elif orig_place.place_type == pt.value:
```
Comment (jehangiramjad) [inline server/lib/nl/fulfillment/base.py:448]: nit: unclear what is the edge case here and why its buggy?

File: static/js/apps/nl_interface/query_result.tsx [hunk 3]
```diff
@@ -167,6 +176,38 @@ export const QueryResult = memo(function QueryResult(
...
+            <div className="nl-query-info">
+              Sorry, there was no relevant data for &quot;
+              {chartsData.placeFallback.origStr}&quot;. &nbsp; Here is data
+              about &quot;{chartsData.placeFallback.newStr}&quot; instead.
+            </div>
+          )}
+          {chartsData && chartsData.placeSource === "PAST_QUERY" && (
+            <div className="nl-query-info">
+              Could not recognize any place in this query, so using{" "}
+              {chartsData.place.name} from a prior query.
```
Comment (jehangiramjad) [inline static/js/apps/nl_interface/query_result.tsx:189]: nit: should this refer to the user's current session? i.e. "... from a prior query in this session"

File: static/js/apps/nl_interface/query_result.tsx [hunk 4]
```diff
@@ -167,6 +176,38 @@ export const QueryResult = memo(function QueryResult(
...
+          {chartsData && chartsData.svSource === "PAST_QUERY" && (
+            <div className="nl-query-info">
+              Could not recognize any topic in this query, so using a topic you
+              previously queried about in this session.
+            </div>
+          )}
+          {chartsData && chartsData.svSource === "UNRECOGNIZED" && (
+            <div className="nl-query-info">
+              Could not recognize any topic from the query. Below are some topic
+              categories with data for {chartsData.place.name}.
```
Comment (jehangiramjad) [inline static/js/apps/nl_interface/query_result.tsx:201]: super nit: should 'data' be replaced with 'observations'?

File: static/js/apps/nl_interface/query_result.tsx [hunk 5]
```diff
@@ -167,6 +176,38 @@ export const QueryResult = memo(function QueryResult(
...
+            </div>
+          )}
+          {chartsData && chartsData.svSource === "UNFULFILLED" && (
+            <div className="nl-query-info">
+              Sorry, there was no relevant data about the topic for{" "}
```
Comment (jehangiramjad) [inline static/js/apps/nl_interface/query_result.tsx:206]: super nit: using 'observations' instead of 'data' sounds better

**Review by jehangiramjad — COMMENTED** [review]
Thanks! There is a lot going on in ``maybe_set_fallback()`` and also the type ``PlaceFallback`` which makes things a bit less clear to me. A couple of other nitty suggestions as well.

**Review by pradh — COMMENTED** [review]
Thanks for the review Jehangir!

**Review by jehangiramjad — APPROVED** [review]
Thanks!

**Pull request merged**